http/src/lib.rs

```diff
@@ -109,40 +100,39 @@ impl<T: hyper::server::Handler<hyper::net::HttpStream> + Send + 'static> From<Op
 /// Allows to intercept request and handle it differently.
 pub trait RequestMiddleware: Send + Sync + 'static {
 	/// Takes a request and decides how to proceed with it.
-	fn on_request(&self, request: &server::Request<hyper::net::HttpStream>) -> RequestMiddlewareAction;
+	fn on_request(&self, request: &server::Request) -> BoxFuture<RequestMiddlewareAction, ()>;
 }
 
 impl<F> RequestMiddleware for F where
-	F: Fn(&server::Request<hyper::net::HttpStream>) -> RequestMiddlewareAction + Sync + Send + 'static,
-{
-	fn on_request(&self, request: &server::Request<hyper::net::HttpStream>) -> RequestMiddlewareAction {
+	F: Fn(&server::Request) -> BoxFuture<RequestMiddlewareAction, ()> + Send + Sync + 'static {
+	fn on_request(&self, request: &server::Request) -> BoxFuture<RequestMiddlewareAction, ()> {
 		(*self)(request)
 	}
 }
 
 #[derive(Default)]
 struct NoopRequestMiddleware;
 impl RequestMiddleware for NoopRequestMiddleware {
-	fn on_request(&self, _request: &server::Request<hyper::net::HttpStream>) -> RequestMiddlewareAction {
-		RequestMiddlewareAction::Proceed {
+	fn on_request(&self, _request: &server::Request) -> BoxFuture<RequestMiddlewareAction, ()> {
+		let action = RequestMiddlewareAction::Proceed {
 			should_continue_on_invalid_cors: false,
-		}
+		};
+		finished(action).boxed()
 	}
 }
 
 /// Extracts metadata from the HTTP request.
 pub trait HttpMetaExtractor<M: jsonrpc::Metadata>: Sync + Send + 'static {
 	/// Read the metadata from the request
-	fn read_metadata(&self, _: &server::Request<hyper::net::HttpStream>) -> M {
+	fn read_metadata(&self, _: &server::Request) -> M {
 		Default::default()
 	}
 }
 
 impl<M, F> HttpMetaExtractor<M> for F where
 	M: jsonrpc::Metadata,
-	F: Fn(&server::Request<hyper::net::HttpStream>) -> M + Sync + Send + 'static,
-{
-	fn read_metadata(&self, req: &server::Request<hyper::net::HttpStream>) -> M {
+	F: Fn(&server::Request) -> M + Sync + Send + 'static {
```

Comment:
Formatting was better before :)

1. No special treatment of last line
2. Visual whitespace between bounds and beginning of implementation